storage_service/common/gpgutils.py

```diff
@@ -0,0 +1,198 @@
+"""GPG Utils.
+
+Contains utilities for interacting with gnupg, i.e., Python GPG. This is
+functionality for listing and creating private/public GPG keys for encrypting
+and decrypting things. It is in its own utility because this functionality is
+needed by both the GPG space and the administration view which can be used to
+manage (i.e., list, created, import, delete) GPG keys.
+
+"""
+
+from __future__ import absolute_import
+# stdlib, alphabetical
+import logging
+
+# Third party dependencies, alphabetical
+import gnupg
+from django.utils.translation import ugettext as _
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+# WARNING: by not providing a gnupghome kw param to ``GPG`` here we are
+# deferring to gnupg to determine where the .gnupg/ dir will be. On a default
+# vagrant/ansible deploy, the .gnupg/ dir will be at
+# /var/lib/archivematica/.gnupg/
+gpg = gnupg.GPG()
+
+
+# Defaults for the default AM SS GPG key
+DFLT_KEY_TYPE = 'RSA'
+DFLT_KEY_LENGTH = 4096
+DFLT_KEY_REAL_NAME = _('Archivematica Storage Service GPG Key')
+DFLT_KEY_PASSPHRASE = ''
+
+
+PASSPHRASED = 'passphrased'
+IMPORT_ERROR = 'import error'
+ENCR_WORKS = 'yes'
+ENCR_FAILS = 'no'
+
+
+def get_gpg_key(fingerprint):
+    """Return the GPG key with fingerprint ``fingerprint`` or None if there is
+    no such key in the SS's GPG keyring.
+    """
+    key_map = gpg.list_keys(True).key_map  # ``True`` means return private keys
+    return key_map.get(fingerprint)
+
+
+def get_gpg_key_list():
+    """Return a list of all GPG keys as dicts. If the Storage Service default
+    key does not exist, we create it here before returning the list.
+    """
+    keys = gpg.list_keys(True)
+    default_key = get_default_gpg_key(keys)
+    if not default_key:
+        generate_default_gpg_key()
+        keys = gpg.list_keys(True)
+    return keys
+
+
+def get_default_gpg_key(keys):
+    """Find the default AM Storage Service key in the existing set of GPG keys
+    or return ``None``.
+    """
+    for key in keys:
+        uids = key['uids']
+        for uid in uids:
+            if uid.startswith(DFLT_KEY_REAL_NAME):
+                return key
+    return None
+
+
+def generate_default_gpg_key():
+    """Generate the default AM Storage Service key. Note that by supplying no
+    expiration date, the key never expires.
+    Warning: generating this key for the first time can take several minutes.
+    At present, this default key is generated the first time that
+    ``get_gpg_key_list`` is called, which is typically when the "Encryption
+    keys" link is clicked under the "Administration" tab. Waiting on a long
+    request like this is not a very good user experience. It may be desirable
+    to do this in a separate thread and poll for completion, or install
+    rng-tools, or else at least warn the user that this will take a long time,
+    or use a smaller key size. The same considerations apply for
+    ``generate_gpg_key``. TODO/QUESTION ^.
```

Comment:
Could we generate the default key in the install/upgrade process?

Reply:
Yes, that is an option. Although what I ended up doing was having Ansible install rng-tools: see https://github.com/artefactual-labs/ansible-archivematica-src/pull/109. With rng-tools installed, GPG key creation time is reduced to the point that this is no longer a usability concern. I think I will leave the code as is and modify the docstring accordingly.

Reply:
Great! Thanks!